lightning/src/routing/router.rs

```diff
@@ -333,84 +577,251 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
+				// At most, we may need the value to be transferred and fees for that transfer.
+				// Assume 900% fees as the highest possible amount we may need, keep in mind fees may be charged on every hop.
+				let mut path_bottleneck_msat = final_value_msat * 10;
+
+				loop {
+					if new_entry.route_hop.short_channel_id == 0 {
+						// Means that the only way to proceed was via a dummy channel we stored for the internal algorithm.
+						// Terminate path finding, because the dummy channel is not usable, and further iterations would only face the exact same issue.
+						break 'path_construction;
+					}
+
+					if let Some(&(_, ref features)) = first_hop_targets.get(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						ordered_hops.last_mut().unwrap().route_hop.node_features = features.to_context();
+					} else if let Some(node) = network.get_nodes().get(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						if let Some(node_info) = node.announcement_info.as_ref() {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = node_info.features.clone();
+						} else {
+							ordered_hops.last_mut().unwrap().route_hop.node_features = NodeFeatures::empty();
+						}
+					} else {
+						// We should be able to fill in features for everything except the last
+						// hop, if the last hop was provided via a BOLT 11 invoice (though we
+						// should be able to extend it further as BOLT 11 does have feature
+						// flags for the last hop node itself).
+						assert!(ordered_hops.last().unwrap().route_hop.pubkey == *payee);
+					}
+
+					if new_entry.available_liquidity_msat > new_entry.min_total_fee_to_payee {
+						// How much value a path can transfer is how much the weakest link can transfer after paying
+						// for the use of the following channels.
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_liquidity_msat - new_entry.min_total_fee_to_payee);
 					} else {
-						res.last_mut().unwrap().node_features = NodeFeatures::empty();
+						// TODO: explain
+						path_bottleneck_msat = cmp::min(path_bottleneck_msat, new_entry.available_liquidity_msat / 10);
 					}
-				} else {
-					// We should be able to fill in features for everything except the last
-					// hop, if the last hop was provided via a BOLT 11 invoice (though we
-					// should be able to extend it further as BOLT 11 does have feature
-					// flags for the last hop node itself).
-					assert!(res.last().unwrap().pubkey == *target);
+
+					if ordered_hops.last().unwrap().route_hop.pubkey == *payee {
+						break;
+					}
+
+					new_entry = match dist.remove(&ordered_hops.last().unwrap().route_hop.pubkey) {
+						Some(payment_hop) => payment_hop,
+						None => {
+							// If we can't reach a given node, something wen't wrong during path traverse.
+							// Stop looking for more paths here because next iterations will face the exact same issue.
+							break 'paths_collection;
```

Comment:
I think the error we have later is sufficient.
Since here I'm looking for more capacity than the payment needs, the payment might actually succeed (`recommended_value_msat = final_value_msat * ROUTE_CAPACITY_PROVISION_FACTOR`). If not, it will be logged later.

Reply:
Ah okay, I agree that it still might succeed but since the comment indicates that this should never happen I thought a log could maybe be useful for debugging if we did hit this case. Not a big deal though, resolving.